### PR TITLE
PERF: is_bool_indexer

### PR DESCRIPTION
**Pull request by jbrockmendel**
Besides avoiding an array allocation, thus avoids the np.asarray call that was one of the main offenders in #41632

### CODE REVIEW

File: pandas/_libs/lib.pyx [hunk 1]
```diff
@@ -2920,3 +2920,41 @@ def to_object_array_tuples(rows: object) -> np.ndarray:
...
+
+    48.1 ns ± 1.26 ns
+    8.1 µs ± 198 ns
+    """
+    cdef:
+        object item
+
+    for item in obj:
+        if not util.is_bool_object(item):
+            return False
```
Comment (WillAyd) [inline pandas/_libs/lib.pyx:2957]: Is there any concern with the benchmarks being somewhat misleading due to short circuiting? At least in the Python space similar functions don't seem to benefit much from this

```python
In [6]: %timeit all(isinstance(x, bool) for x in obj)
6.7 µs ± 136 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

In [7]: %timeit np.array(obj).dtype.kind == "b"
5.98 µs ± 121 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
```

Comment (jbrockmendel) [reply pandas/_libs/lib.pyx:2957]: > Is there any concern with the benchmarks being somewhat misleading due to short circuiting?

Only one of the 4 cases posted short-circuits.  Am I misunderstanding the question?

>  At least in the Python space similar functions don't seem to benefit much from this

I was surprised by this too, found that doing it in cython was much more performant after talking with @seberg yesterday about how numpy does this inference.

Comment (seberg) [reply pandas/_libs/lib.pyx:2957]: Wow, the difference is big...

NumPy should maybe try to be a bit faster, but that will just make the gap a bit smaller (e.g. by assuming that usually all elements are the same type, I guess a true casting support `np.asarray(..., dtype=bool, casting="equiv")` would fill the gap, but no idea how likely that would be).  Unlike NumPy, you already know what you want here (bools) and what you got (a 1-d list).

Comment (WillAyd) [reply pandas/_libs/lib.pyx:2957]: Ah my bad on short circuiting - misread it. Thanks for comments

**Review by WillAyd — APPROVED** [review]

**Comment by jreback** [comment]
thanks this looks good

**Pull request merged**